### PR TITLE
gRPC: Implement max_receive_message_length in envoy gRPC

### PR DESCRIPTION
**Pull request by tyxia**
Over-frame-limit error can now be surfaced from gRPC codec level after PR #32511

Envoy gRPC client can leverage this to implement max_recevie_meassage_length with config field. This field is defaulted to 0, which means no limits.
-  When the message is over limit, it can be rejected  before frame data [is fully decoded (i.e. expanded](https://github.com/envoyproxy/envoy/blob/ffbf1889ef788ccb4b1d0e402ed2c3976d7204fd/source/common/grpc/async_client_impl.cc#L144)). 
- This can prevent malicious attack , for example, unbounded and huge message is sent over channel and is injected and buffered in Envoy over Envoy-gRPC.

error status will be improved by https://github.com/envoyproxy/envoy/pull/32676

### CODE REVIEW

File: api/envoy/config/core/v3/grpc_service.proto [hunk 1]
```diff
@@ -49,6 +49,10 @@ message GrpcService {
     // Currently only supported for xDS gRPC streams.
     // If not set, xDS gRPC streams default base interval:500ms, maximum interval:30s will be applied.
     RetryPolicy retry_policy = 3;
+
+    // Maximum receive message length.
+    // Defaults to 0, which means unlimited.
+    uint32 max_receive_message_length = 4;
```
Comment (htuch) [inline api/envoy/config/core/v3/grpc_service.proto:55]: As per internal thread, this should be WKT.

Comment (tyxia) [reply api/envoy/config/core/v3/grpc_service.proto]: Good idea! Yes, this provides the flexibility in OSS that we may want to default to safe limited length in the future.

Done.

File: source/common/grpc/codec.cc [hunk 2]
```diff
@@ -38,7 +38,6 @@ bool Decoder::decode(Buffer::Instance& input, std::vector<Frame>& output) {
   output_ = &output;
   inspect(input);
   output_ = nullptr;
-
```
Comment (htuch) [inline source/common/grpc/codec.cc:41]: Nit: unrelated?

**Review by htuch — DISMISSED** [review]
LGTM

File: source/common/grpc/codec.cc (hunk 2 above)
Comment (tyxia) [reply source/common/grpc/codec.cc:41]: Thanks!

I actually revert this unrelated change in next pr https://github.com/envoyproxy/envoy/pull/32676.  Can I change it there?(just want to avoid the long CI run if it is ok)

Comment (htuch) [reply source/common/grpc/codec.cc:41]: Sure

Comment (htuch) [reply source/common/grpc/codec.cc:41]: When do you plan on adding release notes?

Comment (tyxia) [reply source/common/grpc/codec.cc:41]: Good point. Can I also add it in https://github.com/envoyproxy/envoy/pull/32676 which is for gRPC error code improvement? because i am thinking that is where I complete this effort.

If you prefer rel note to be added here, i am also happy to do so.

File: api/envoy/config/core/v3/grpc_service.proto [hunk 3]
```diff
@@ -49,6 +49,10 @@ message GrpcService {
     // Currently only supported for xDS gRPC streams.
     // If not set, xDS gRPC streams default base interval:500ms, maximum interval:30s will be applied.
     RetryPolicy retry_policy = 3;
+
+    // Maximum receive message length.
```
Comment (yanavlasov) [inline api/envoy/config/core/v3/grpc_service.proto:53]:
```suggestion
    // Maximum gRPC  message size that is allowed to be received. This limit is applied to individual messages in the streaming response and not the total size of streaming response.
    // If a message over this limit is received the gRPC request is terminated with the  RESOURCE_EXHAUSTED error.
```

**Review by yanavlasov — CHANGES_REQUESTED** [review]
LGTM, but please clarify the behavior of the new option in the comment.

File: api/envoy/config/core/v3/grpc_service.proto (hunk 3 above)
Comment (yanavlasov) [reply api/envoy/config/core/v3/grpc_service.proto]: Streaming vs unary response and size limitations are not a client detail. If I'm a developer I need to know how this limit is applied in the unary and streaming gRPC responses. While unary is self-explanatory, since there is only one message, the streaming case is unclear. Does this limit apply to the entire streamed response or just the individual messages within the stream? Clarification would be very helpful.

Comment (tyxia) [reply api/envoy/config/core/v3/grpc_service.proto]: fair point. Thanks! Done.

**Comment by yanavlasov** [comment]
Will wait for Lizan's comment to be addressed.

/wait

**Comment by tyxia** [comment]
Sorry for delay. Release note has been added.

PTAL @htuch @yanavlasov @lizan

**Comment by soulxu** [comment]
gentle ping @htuch @yanavlasov @lizan

**Review by htuch — APPROVED** [review]

**Review by lizan — APPROVED** [review]

**Comment by tyxia** [comment]
gentle ping on merge (probably @soulxu ?)

I think all the reviewers have approved this PR. Thanks!

**Review by yanavlasov — APPROVED** [review]
